[WebGPU] Align SurfaceDescriptor and SwapChainDescriptor with the spec and WebGPU.h

litherum opened the pull request:
#### ffc98ceb72ee397bd83b6d7d711fcb760b712c34
<pre>
[WebGPU] Align SurfaceDescriptor and SwapChainDescriptor with the spec and WebGPU.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=250994">https://bugs.webkit.org/show_bug.cgi?id=250994</a>
rdar://104539553

Reviewed by Dean Jackson.

Conceptually, a Surface corresponds to a GPUCanvasContext, and a SwapChain corresponds to a call
to GPUCanvasContext.configure(). Both the spec and WebGPU.h agree that the Surface should be
constructable with almost nothing - you don&apos;t even need a device or any WebGPU objects to create
a GPUCanvasContext object from a canvas. Creating a swap chain (via GPUCanvasContext.configure())
is where the guts are - that&apos;s where all the properties of the textures are specified, so that&apos;s
where we can finally create IOSurfaces.

Right now, we treat Surfaces and SwapChains as siblings - they have the same lifetime inside
GPUCanvasContext.cpp, so this patch is purely mechanical right now. However, there will be 2
follow-up patches in the future:

1. Connecting CompositorIntegration (from <a href="https://github.com/WebKit/WebKit/pull/8933)">https://github.com/WebKit/WebKit/pull/8933)</a> to Surface.
       This will provide a natural owner for the IOSurface renderbuffers, as well as a message
       receiver so these IOSurfaces can be communicated to the compositor. Mechanically, this
       change will add a CompositorIntegration inside SurfaceDescriptor.
2. We can move the lifetime of Surface to be created a bit earlier than SwapChain, because there&apos;s
       no reason to delay creating it; it can be created with almost no information.

A special note about width and height: The WebGPU spec used to include these as arguments to
GPUCanvasContext.configure(), but we changed the spec to have that information come from the
canvas itself rather than being supplied by the site author. So, conceptually, these fields
still belong in the SwapChainDescriptor, and they are present there in this patch. This also
matches what WebGPU.h says - those fields are present in the SwapChainDescriptor.

* Source/WebCore/Modules/WebGPU/GPUSurfaceDescriptor.h:
(WebCore::GPUSurfaceDescriptor::convertToBacking const):
(): Deleted.
* Source/WebCore/Modules/WebGPU/GPUSwapChainDescriptor.h:
(WebCore::GPUSwapChainDescriptor::convertToBacking const):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createSurface):
(PAL::WebGPU::DeviceImpl::createSwapChain):
(PAL::WebGPU::convertToWidthHeight): Deleted.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSurfaceDescriptor.h:
(): Deleted.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSwapChainDescriptor.h:
* Source/WebCore/html/canvas/GPUCanvasContext.cpp:
(WebCore::GPUCanvasContext::createSwapChainIfNeeded):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::createIOSurface):
(WebGPU::createSurfaceFromDescriptor):
(WebGPU::PresentationContextIOSurface::PresentationContextIOSurface):
(WebGPU::PresentationContextIOSurface::configure):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.h:
(): Deleted.
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.serialization.in:
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.serialization.in:

Canonical link: <a href="https://commits.webkit.org/259312@main">https://commits.webkit.org/259312@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/882ca2750c50a9c5ccba849204f7768839a1f53a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104469 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13547 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37374 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113747 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173971 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108390 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14652 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4466 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96811 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112714 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110236 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11295 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94349 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38891 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93160 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25959 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80560 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6905 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27316 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7031 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3877 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13062 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46876 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6421 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8825 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (commented):
~~Is `addCompletedHandler` the appropriate sync point to wait for to avoid the race here?~~

~~Can we really call `-[CALayer setContents:machPort]` when the command buffers are submitted but not completed without encountering the same expected race condition?~~

mwyrzykowski commented on Source/WebCore/html/canvas/GPUCanvasContext.cpp:
```diff
@@ -116,21 +116,21 @@ void GPUCanvasContext::createSwapChainIfNeeded()
+        m_configuration->viewFormats,
+        m_configuration->colorSpace,
+        m_configuration->compositingAlphaMode,
+        static_cast<uint32_t>(m_width), // FIXME: Is it possible for these to be negative?
```
Isn't it not recommended in C++ to use unsigned types simply because the value is only expected to be positive?

The C++ core guidelines recommend against using [unsigned to avoid negative values](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-nonnegative), Section 4.4 on page 73 of The C++ Programming Language says unsigned types should be used for bitfields and not in an attempt to ensure values are positive, other talks by people on the C++ standards committee (e.g., Herb Sutter) recommend against using unsigned types simply because the value is expected to by positive. Plus I think it was one of the reasons for std::ssize in C++20

Do we have official guidance in WebKit? If not, I think it would make more sense to preferred signed types for values which are not bitfields.

litherum replied:
Oh, this is the first I've heard about this recommendation. I'm just using unsigneds here because that's what WebGPU.h uses: https://github.com/webgpu-native/webgpu-headers/blob/main/webgpu.h#L975

So the cast is going to happen at _some_ point, either here or in `DeviceImpl::createSwapChain()`. Do you have a preference where the cast happens?

(I also disagree with those C++ guidelines, but that's a discussion for another time...)

mwyrzykowski replied:
I don’t have any strong opinions since it’s unlikely we’ll perform arithmetic or loops on these integers, it would be fine for me to be uint32_t in this case

litherum replied:
🆗

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/GPUSwapChainDescriptor.h:
```diff
@@ -39,17 +41,25 @@ struct GPUSwapChainDescriptor : public GPUObjectDescriptorBase {
 
-    GPUExtent3D size;
-    GPUSize32 sampleCount { 1 };
     GPUTextureFormat format { GPUTextureFormat::R8unorm };
```
should we make this `GPUTextureFormat::BGR8unorm` while we are here or does it not really matter since its a required field from the web api?

litherum replied:
I don't think it really matters, but I'm happy to change it now.

djg commented on Source/WebCore/Modules/WebGPU/GPUSwapChainDescriptor.h:
```diff
@@ -39,17 +41,25 @@ struct GPUSwapChainDescriptor : public GPUObjectDescriptorBase {
     {
         return {
             { label },
-            WebCore::convertToBacking(size),
-            sampleCount,
             WebCore::convertToBacking(format),
             convertTextureUsageFlagsToBacking(usage),
+            viewFormats.map([] (auto& viewFormat) {
+                return WebCore::convertToBacking(viewFormat);
+            }),
+            WebCore::convertToBacking(colorSpace),
+            WebCore::convertToBacking(compositingAlphaMode),
+            width,
+            height,
         };
```
Can we move implementation into `.cpp`/`.mm` as much as possible?

litherum replied:
None of these Descriptor.h files have accompanying .cpp files, mostly just for my own sanity when dealing with all this stuff. If we're going to start giving Descriptors accompanying .cpp files, we should do it consistently for all of them, which is a pretty substantial change I think, and would probably best be done in another patch (probably after some additional discussion among the team, too)

djg reviewed (approved):
:shipit:

grorg reviewed (approved):
rs=me based on previous patches

litherum committed:
[WebGPU] Align SurfaceDescriptor and SwapChainDescriptor with the spec and WebGPU.h
https://bugs.webkit.org/show_bug.cgi?id=250994
rdar://104539553

Reviewed by Dean Jackson.

Conceptually, a Surface corresponds to a GPUCanvasContext, and a SwapChain corresponds to a call
to GPUCanvasContext.configure(). Both the spec and WebGPU.h agree that the Surface should be
constructable with almost nothing - you don't even need a device or any WebGPU objects to create
a GPUCanvasContext object from a canvas. Creating a swap chain (via GPUCanvasContext.configure())
is where the guts are - that's where all the properties of the textures are specified, so that's
where we can finally create IOSurfaces.

Right now, we treat Surfaces and SwapChains as siblings - they have the same lifetime inside
GPUCanvasContext.cpp, so this patch is purely mechanical right now. However, there will be 2
follow-up patches in the future:

1. Connecting CompositorIntegration (from https://github.com/WebKit/WebKit/pull/8933) to Surface.
       This will provide a natural owner for the IOSurface renderbuffers, as well as a message
       receiver so these IOSurfaces can be communicated to the compositor. Mechanically, this
       change will add a CompositorIntegration inside SurfaceDescriptor.
2. We can move the lifetime of Surface to be created a bit earlier than SwapChain, because there's
       no reason to delay creating it; it can be created with almost no information.

A special note about width and height: The WebGPU spec used to include these as arguments to
GPUCanvasContext.configure(), but we changed the spec to have that information come from the
canvas itself rather than being supplied by the site author. So, conceptually, these fields
still belong in the SwapChainDescriptor, and they are present there in this patch. This also
matches what WebGPU.h says - those fields are present in the SwapChainDescriptor.

* Source/WebCore/Modules/WebGPU/GPUSurfaceDescriptor.h:
(WebCore::GPUSurfaceDescriptor::convertToBacking const):
(): Deleted.
* Source/WebCore/Modules/WebGPU/GPUSwapChainDescriptor.h:
(WebCore::GPUSwapChainDescriptor::convertToBacking const):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createSurface):
(PAL::WebGPU::DeviceImpl::createSwapChain):
(PAL::WebGPU::convertToWidthHeight): Deleted.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSurfaceDescriptor.h:
(): Deleted.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUSwapChainDescriptor.h:
* Source/WebCore/html/canvas/GPUCanvasContext.cpp:
(WebCore::GPUCanvasContext::createSwapChainIfNeeded):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::createIOSurface):
(WebGPU::createSurfaceFromDescriptor):
(WebGPU::PresentationContextIOSurface::PresentationContextIOSurface):
(WebGPU::PresentationContextIOSurface::configure):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.h:
(): Deleted.
* Source/WebKit/Shared/WebGPU/WebGPUSurfaceDescriptor.serialization.in:
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPUSwapChainDescriptor.serialization.in:

Canonical link: https://commits.webkit.org/259312@main
